// immutable-json-patch@6.0.1/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  immutableJSONPatch: true,
  parsePath: true,
  parseFrom: true,
  revertJSONPatch: true,
  getIn: true,
  setIn: true,
  existsIn: true,
  insertAt: true,
  deleteIn: true,
  updateIn: true,
  transform: true
};
Object.defineProperty(exports, "deleteIn", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.deleteIn;
  }
});
Object.defineProperty(exports, "existsIn", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.existsIn;
  }
});
Object.defineProperty(exports, "getIn", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.getIn;
  }
});
Object.defineProperty(exports, "immutableJSONPatch", {
  enumerable: true,
  get: function () {
    return _immutableJSONPatch.immutableJSONPatch;
  }
});
Object.defineProperty(exports, "insertAt", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.insertAt;
  }
});
Object.defineProperty(exports, "parseFrom", {
  enumerable: true,
  get: function () {
    return _immutableJSONPatch.parseFrom;
  }
});
Object.defineProperty(exports, "parsePath", {
  enumerable: true,
  get: function () {
    return _immutableJSONPatch.parsePath;
  }
});
Object.defineProperty(exports, "revertJSONPatch", {
  enumerable: true,
  get: function () {
    return _revertJSONPatch.revertJSONPatch;
  }
});
Object.defineProperty(exports, "setIn", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.setIn;
  }
});
Object.defineProperty(exports, "transform", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.transform;
  }
});
Object.defineProperty(exports, "updateIn", {
  enumerable: true,
  get: function () {
    return _immutabilityHelpers.updateIn;
  }
});
var _immutableJSONPatch = require("./immutableJSONPatch.js");
var _revertJSONPatch = require("./revertJSONPatch.js");
var _types = require("./types.js");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _jsonPointer = require("./jsonPointer.js");
Object.keys(_jsonPointer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _jsonPointer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _jsonPointer[key];
    }
  });
});
var _typeguards = require("./typeguards.js");
Object.keys(_typeguards).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _typeguards[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _typeguards[key];
    }
  });
});
var _immutabilityHelpers = require("./immutabilityHelpers.js");
