// @tanstack/eslint-plugin-query@5.66.0/dist/cjs/rules/infinite-query-property-order/infinite-query-property-order.rule.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@typescript-eslint/utils");
const getDocsUrl = require("../../utils/get-docs-url.cjs");
const detectReactQueryImports = require("../../utils/detect-react-query-imports.cjs");
const infiniteQueryPropertyOrder_utils = require("./infinite-query-property-order.utils.cjs");
const constants = require("./constants.cjs");
const createRule = utils.ESLintUtils.RuleCreator(getDocsUrl.getDocsUrl);
const infiniteQueryFunctionsSet = new Set(constants.infiniteQueryFunctions);
function isInfiniteQueryFunction(node) {
  return infiniteQueryFunctionsSet.has(node);
}
const name = "infinite-query-property-order";
const rule = createRule({
  name,
  meta: {
    type: "problem",
    docs: {
      description: "Ensure correct order of inference sensitive properties for infinite queries",
      recommended: "error"
    },
    messages: {
      invalidOrder: "Invalid order of properties for `{{function}}`."
    },
    schema: [],
    hasSuggestions: true,
    fixable: "code"
  },
  defaultOptions: [],
  create: detectReactQueryImports.detectTanstackQueryImports((context) => {
    return {
      CallExpression(node) {
        if (node.callee.type !== utils.AST_NODE_TYPES.Identifier) {
          return;
        }
        const infiniteQueryFunction = node.callee.name;
        if (!isInfiniteQueryFunction(infiniteQueryFunction)) {
          return;
        }
        const argument = node.arguments[0];
        if (argument === void 0 || argument.type !== "ObjectExpression") {
          return;
        }
        const allProperties = argument.properties;
        if (allProperties.length < 2) {
          return;
        }
        const properties = allProperties.flatMap((p, index) => {
          if (p.type === utils.AST_NODE_TYPES.Property && p.key.type === utils.AST_NODE_TYPES.Identifier) {
            return { name: p.key.name, property: p };
          } else return { name: `_property_${index}`, property: p };
        });
        const sortedProperties = infiniteQueryPropertyOrder_utils.sortDataByOrder(properties, constants.sortRules, "name");
        if (sortedProperties === null) {
          return;
        }
        context.report({
          node: argument,
          data: { function: node.callee.name },
          messageId: "invalidOrder",
          fix(fixer) {
            const sourceCode = context.sourceCode;
            const reorderedText = sortedProperties.reduce(
              (sourceText, specifier, index) => {
                let textBetweenProperties = "";
                if (index < allProperties.length - 1) {
                  textBetweenProperties = sourceCode.getText().slice(
                    allProperties[index].range[1],
                    allProperties[index + 1].range[0]
                  );
                }
                return sourceText + sourceCode.getText(specifier.property) + textBetweenProperties;
              },
              ""
            );
            return fixer.replaceTextRange(
              [allProperties[0].range[0], allProperties.at(-1).range[1]],
              reorderedText
            );
          }
        });
      }
    };
  })
});
exports.name = name;
exports.rule = rule;
