// cypress-terminal-report@7.1.0/src/installLogsPrinter.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallLogsPrinterSchema = void 0;
const constants_1 = __importDefault(require("./constants"));
const superstruct_1 = require("superstruct");
const LOG_OCCURRENCE = Object.values(constants_1.default.LOG_OCCURRENCE);
const InstallLogsPrinterSchema = (0, superstruct_1.object)({
    printLogsToConsole: (0, superstruct_1.optional)((0, superstruct_1.enums)(LOG_OCCURRENCE)),
    printLogsToFile: (0, superstruct_1.optional)((0, superstruct_1.enums)(LOG_OCCURRENCE)),
    includeSuccessfulHookLogs: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    defaultTrimLength: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    commandTrimLength: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    routeTrimLength: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    compactLogs: (0, superstruct_1.optional)((0, superstruct_1.min)((0, superstruct_1.number)(), 0)),
    outputCompactLogs: (0, superstruct_1.optional)((0, superstruct_1.union)([(0, superstruct_1.min)((0, superstruct_1.number)(), -1), (0, superstruct_1.boolean)()])),
    outputRoot: (0, superstruct_1.optional)((0, superstruct_1.string)()),
    specRoot: (0, superstruct_1.optional)((0, superstruct_1.string)()),
    outputTarget: (0, superstruct_1.optional)((0, superstruct_1.record)((0, superstruct_1.string)(), (0, superstruct_1.union)([(0, superstruct_1.string)(), (0, superstruct_1.func)()]))),
    collectTestLogs: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    logToFilesOnAfterRun: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    outputVerbose: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    debug: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
});
exports.InstallLogsPrinterSchema = InstallLogsPrinterSchema;
