// cypress-terminal-report@7.1.0/src/installLogsCollector.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallLogsCollectorSchema = void 0;
const constants_1 = __importDefault(require("./constants"));
const superstruct_1 = require("superstruct");
const InstallLogsCollectorSchema = (0, superstruct_1.object)({
    collectTypes: (0, superstruct_1.optional)((0, superstruct_1.array)((0, superstruct_1.enums)(Object.values(constants_1.default.LOG_TYPES).filter((t) => t !== constants_1.default.LOG_TYPES.PLUGIN_LOG_TYPE)))),
    filterLog: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    processLog: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    collectTestLogs: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    xhr: (0, superstruct_1.optional)((0, superstruct_1.object)({
        printHeaderData: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
        printRequestData: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
        printBody: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    })),
    enableExtendedCollector: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    enableContinuousLogging: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    commandTimings: (0, superstruct_1.optional)((0, superstruct_1.enums)(Object.values(constants_1.default.COMMAND_TIMINGS))),
    debug: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
});
exports.InstallLogsCollectorSchema = InstallLogsCollectorSchema;
