// @rjsf/chakra-ui@5.23.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Form: () => Form_default,
  Templates: () => Templates_default,
  Theme: () => Theme_default,
  Widgets: () => Widgets_default,
  __createChakraFrameProvider: () => __createChakraFrameProvider,
  default: () => src_default,
  generateForm: () => generateForm,
  generateTemplates: () => generateTemplates,
  generateTheme: () => generateTheme,
  generateWidgets: () => generateWidgets
});
module.exports = __toCommonJS(src_exports);

// src/Form/Form.tsx
var import_core = require("@rjsf/core");

// src/AddButton/AddButton.tsx
var import_utils = require("@rjsf/utils");
var import_react = require("@chakra-ui/react");
var import_icons = require("@chakra-ui/icons");
var import_jsx_runtime = require("react/jsx-runtime");
function AddButton({
  uiSchema,
  registry,
  ...props
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Button, { leftIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icons.AddIcon, {}), ...props, children: translateString(import_utils.TranslatableString.AddItemButton) });
}

// src/ArrayFieldItemTemplate/ArrayFieldItemTemplate.tsx
var import_react2 = require("react");
var import_react3 = require("@chakra-ui/react");
var import_jsx_runtime2 = require("react/jsx-runtime");
function ArrayFieldItemTemplate(props) {
  const {
    children,
    disabled,
    hasToolbar,
    hasCopy,
    hasMoveDown,
    hasMoveUp,
    hasRemove,
    index,
    onCopyIndexClick,
    onDropIndexClick,
    onReorderClick,
    readonly,
    uiSchema,
    registry
  } = props;
  const { CopyButton: CopyButton2, MoveDownButton: MoveDownButton2, MoveUpButton: MoveUpButton2, RemoveButton: RemoveButton2 } = registry.templates.ButtonTemplates;
  const onCopyClick = (0, import_react2.useMemo)(() => onCopyIndexClick(index), [index, onCopyIndexClick]);
  const onRemoveClick = (0, import_react2.useMemo)(() => onDropIndexClick(index), [index, onDropIndexClick]);
  const onArrowUpClick = (0, import_react2.useMemo)(() => onReorderClick(index, index - 1), [index, onReorderClick]);
  const onArrowDownClick = (0, import_react2.useMemo)(() => onReorderClick(index, index + 1), [index, onReorderClick]);
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(import_react3.HStack, { alignItems: "flex-end", py: 1, children: [
    /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_react3.Box, { w: "100%", children }),
    hasToolbar && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_react3.Box, { children: /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(import_react3.ButtonGroup, { isAttached: true, mb: 1, children: [
      (hasMoveUp || hasMoveDown) && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        MoveUpButton2,
        {
          disabled: disabled || readonly || !hasMoveUp,
          onClick: onArrowUpClick,
          uiSchema,
          registry
        }
      ),
      (hasMoveUp || hasMoveDown) && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        MoveDownButton2,
        {
          disabled: disabled || readonly || !hasMoveDown,
          onClick: onArrowDownClick,
          uiSchema,
          registry
        }
      ),
      hasCopy && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        CopyButton2,
        {
          disabled: disabled || readonly,
          onClick: onCopyClick,
          uiSchema,
          registry
        }
      ),
      hasRemove && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        RemoveButton2,
        {
          disabled: disabled || readonly,
          onClick: onRemoveClick,
          uiSchema,
          registry
        }
      )
    ] }) })
  ] });
}

// src/ArrayFieldTemplate/ArrayFieldTemplate.tsx
var import_react4 = require("@chakra-ui/react");
var import_utils2 = require("@rjsf/utils");
var import_jsx_runtime3 = require("react/jsx-runtime");
function ArrayFieldTemplate(props) {
  const { canAdd, disabled, idSchema, uiSchema, items, onAddClick, readonly, registry, required, schema, title } = props;
  const uiOptions = (0, import_utils2.getUiOptions)(uiSchema);
  const ArrayFieldDescriptionTemplate = (0, import_utils2.getTemplate)(
    "ArrayFieldDescriptionTemplate",
    registry,
    uiOptions
  );
  const ArrayFieldItemTemplate2 = (0, import_utils2.getTemplate)(
    "ArrayFieldItemTemplate",
    registry,
    uiOptions
  );
  const ArrayFieldTitleTemplate = (0, import_utils2.getTemplate)(
    "ArrayFieldTitleTemplate",
    registry,
    uiOptions
  );
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_react4.Box, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      ArrayFieldTitleTemplate,
      {
        idSchema,
        title: uiOptions.title || title,
        schema,
        uiSchema,
        required,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      ArrayFieldDescriptionTemplate,
      {
        idSchema,
        description: uiOptions.description || schema.description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_react4.Grid, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_react4.GridItem, { children: items.length > 0 && items.map(({ key, ...itemProps }) => /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(ArrayFieldItemTemplate2, { ...itemProps }, key)) }),
      canAdd && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_react4.GridItem, { justifySelf: "flex-end", children: /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_react4.Box, { mt: 2, children: /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
        AddButton2,
        {
          className: "array-item-add",
          onClick: onAddClick,
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      ) }) })
    ] }, `array-item-list-${idSchema.$id}`)
  ] });
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
var import_react6 = require("@chakra-ui/react");
var import_utils3 = require("@rjsf/utils");

// src/utils.ts
var import_react5 = require("@chakra-ui/react");
function getChakra({ uiSchema = {} }) {
  const chakraProps = uiSchema["ui:options"] && uiSchema["ui:options"].chakra || {};
  Object.keys(chakraProps).forEach((key) => {
    if ((0, import_react5.shouldForwardProp)(key)) {
      delete chakraProps[key];
    }
  });
  return chakraProps;
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
var import_jsx_runtime4 = require("react/jsx-runtime");
function BaseInputTemplate(props) {
  const {
    id,
    type,
    value,
    label,
    hideLabel,
    schema,
    uiSchema,
    onChange,
    onChangeOverride,
    onBlur,
    onFocus,
    options,
    required,
    readonly,
    rawErrors,
    autofocus,
    placeholder,
    disabled
  } = props;
  const inputProps = (0, import_utils3.getInputProps)(schema, type, options);
  const chakraProps = getChakra({ uiSchema });
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)(
    import_react6.FormControl,
    {
      mb: 1,
      ...chakraProps,
      isDisabled: disabled || readonly,
      isRequired: required,
      isReadOnly: readonly,
      isInvalid: rawErrors && rawErrors.length > 0,
      children: [
        (0, import_utils3.labelValue)(
          /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(import_react6.FormLabel, { htmlFor: id, id: `${id}-label`, children: label }),
          hideLabel || !label
        ),
        /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
          import_react6.Input,
          {
            id,
            name: id,
            value: value || value === 0 ? value : "",
            onChange: onChangeOverride || _onChange,
            onBlur: _onBlur,
            onFocus: _onFocus,
            autoFocus: autofocus,
            placeholder,
            ...inputProps,
            list: schema.examples ? (0, import_utils3.examplesId)(id) : void 0,
            "aria-describedby": (0, import_utils3.ariaDescribedByIds)(id, !!schema.examples)
          }
        ),
        Array.isArray(schema.examples) ? /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("datalist", { id: (0, import_utils3.examplesId)(id), children: schema.examples.concat(schema.default && !schema.examples.includes(schema.default) ? [schema.default] : []).map((example) => {
          return /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("option", { value: example }, example);
        }) }) : null
      ]
    }
  );
}

// src/DescriptionField/DescriptionField.tsx
var import_react7 = require("@chakra-ui/react");
var import_jsx_runtime5 = require("react/jsx-runtime");
function DescriptionField({ description, id }) {
  if (!description) {
    return null;
  }
  if (typeof description === "string") {
    return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_react7.Text, { as: "sup", fontSize: "md", id, children: description });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_jsx_runtime5.Fragment, { children: description });
}

// src/ErrorList/ErrorList.tsx
var import_utils5 = require("@rjsf/utils");
var import_react8 = require("@chakra-ui/react");
var import_icons2 = require("@chakra-ui/icons");
var import_jsx_runtime6 = require("react/jsx-runtime");
function ErrorList({
  errors,
  registry
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(import_react8.Alert, { flexDirection: "column", alignItems: "flex-start", gap: 3, status: "error", children: [
    /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(import_react8.AlertTitle, { children: translateString(import_utils5.TranslatableString.ErrorsLabel) }),
    /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(import_react8.List, { children: errors.map((error, i) => /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(import_react8.ListItem, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(import_react8.ListIcon, { as: import_icons2.WarningIcon, color: "red.500" }),
      error.stack
    ] }, i)) })
  ] });
}

// src/IconButton/ChakraIconButton.tsx
var import_react9 = require("react");
var import_react10 = require("@chakra-ui/react");
var import_jsx_runtime7 = require("react/jsx-runtime");
function ChakraIconButton(props) {
  const { icon, iconType, uiSchema, registry, ...otherProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(import_react10.IconButton, { "aria-label": props.title, ...otherProps, icon });
}
ChakraIconButton.displayName = "ChakraIconButton";
var ChakraIconButton_default = (0, import_react9.memo)(ChakraIconButton);

// src/IconButton/IconButton.tsx
var import_utils6 = require("@rjsf/utils");
var import_icons3 = require("@chakra-ui/icons");
var import_jsx_runtime8 = require("react/jsx-runtime");
function CopyButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(ChakraIconButton_default, { title: translateString(import_utils6.TranslatableString.CopyButton), ...props, icon: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(import_icons3.CopyIcon, {}) });
}
function MoveDownButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
    ChakraIconButton_default,
    {
      title: translateString(import_utils6.TranslatableString.MoveDownButton),
      ...props,
      icon: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(import_icons3.ArrowDownIcon, {})
    }
  );
}
function MoveUpButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
    ChakraIconButton_default,
    {
      title: translateString(import_utils6.TranslatableString.MoveUpButton),
      ...props,
      icon: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(import_icons3.ArrowUpIcon, {})
    }
  );
}
function RemoveButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
    ChakraIconButton_default,
    {
      title: translateString(import_utils6.TranslatableString.RemoveButton),
      ...props,
      icon: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(import_icons3.DeleteIcon, {})
    }
  );
}

// src/FieldErrorTemplate/FieldErrorTemplate.tsx
var import_utils7 = require("@rjsf/utils");
var import_react11 = require("@chakra-ui/react");
var import_jsx_runtime9 = require("react/jsx-runtime");
function FieldErrorTemplate(props) {
  const { errors = [], idSchema } = props;
  if (errors.length === 0) {
    return null;
  }
  const id = (0, import_utils7.errorId)(idSchema);
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(import_react11.List, { children: errors.map((error, i) => {
    return /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(import_react11.ListItem, { children: /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(import_react11.FormErrorMessage, { id, children: error }) }, i);
  }) });
}

// src/FieldHelpTemplate/FieldHelpTemplate.tsx
var import_utils8 = require("@rjsf/utils");
var import_react12 = require("@chakra-ui/react");
var import_jsx_runtime10 = require("react/jsx-runtime");
function FieldHelpTemplate(props) {
  const { idSchema, help } = props;
  if (!help) {
    return null;
  }
  const id = (0, import_utils8.helpId)(idSchema);
  return /* @__PURE__ */ (0, import_jsx_runtime10.jsx)(import_react12.FormHelperText, { id, children: help });
}

// src/FieldTemplate/FieldTemplate.tsx
var import_utils9 = require("@rjsf/utils");
var import_react13 = require("@chakra-ui/react");
var import_jsx_runtime11 = require("react/jsx-runtime");
function FieldTemplate(props) {
  const {
    id,
    children,
    classNames,
    style,
    disabled,
    displayLabel,
    hidden,
    label,
    onDropPropertyClick,
    onKeyChange,
    readonly,
    registry,
    required,
    rawErrors = [],
    errors,
    help,
    description,
    rawDescription,
    schema,
    uiSchema
  } = props;
  const uiOptions = (0, import_utils9.getUiOptions)(uiSchema);
  const WrapIfAdditionalTemplate2 = (0, import_utils9.getTemplate)(
    "WrapIfAdditionalTemplate",
    registry,
    uiOptions
  );
  if (hidden) {
    return /* @__PURE__ */ (0, import_jsx_runtime11.jsx)("div", { style: { display: "none" }, children });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(
    WrapIfAdditionalTemplate2,
    {
      classNames,
      style,
      disabled,
      id,
      label,
      onDropPropertyClick,
      onKeyChange,
      readonly,
      required,
      schema,
      uiSchema,
      registry,
      children: /* @__PURE__ */ (0, import_jsx_runtime11.jsxs)(import_react13.FormControl, { isRequired: required, isInvalid: rawErrors && rawErrors.length > 0, children: [
        children,
        displayLabel && rawDescription ? /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(import_react13.Text, { mt: 2, children: description }) : null,
        errors,
        help
      ] })
    }
  );
}

// src/ObjectFieldTemplate/ObjectFieldTemplate.tsx
var import_react14 = require("@chakra-ui/react");
var import_utils10 = require("@rjsf/utils");
var import_jsx_runtime12 = require("react/jsx-runtime");
function ObjectFieldTemplate(props) {
  const {
    description,
    title,
    properties,
    required,
    disabled,
    readonly,
    uiSchema,
    idSchema,
    schema,
    formData,
    onAddClick,
    registry
  } = props;
  const uiOptions = (0, import_utils10.getUiOptions)(uiSchema);
  const TitleFieldTemplate = (0, import_utils10.getTemplate)("TitleFieldTemplate", registry, uiOptions);
  const DescriptionFieldTemplate = (0, import_utils10.getTemplate)(
    "DescriptionFieldTemplate",
    registry,
    uiOptions
  );
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ (0, import_jsx_runtime12.jsxs)(import_jsx_runtime12.Fragment, { children: [
    title && /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      TitleFieldTemplate,
      {
        id: (0, import_utils10.titleId)(idSchema),
        title,
        required,
        schema,
        uiSchema,
        registry
      }
    ),
    description && /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
      DescriptionFieldTemplate,
      {
        id: (0, import_utils10.descriptionId)(idSchema),
        description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime12.jsxs)(import_react14.Grid, { gap: description ? 2 : 6, mb: 4, children: [
      properties.map(
        (element, index) => element.hidden ? element.content : /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(import_react14.GridItem, { children: element.content }, `${idSchema.$id}-${element.name}-${index}`)
      ),
      (0, import_utils10.canExpand)(schema, uiSchema, formData) && /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(import_react14.GridItem, { justifySelf: "flex-end", children: /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
        AddButton2,
        {
          className: "object-property-expand",
          onClick: onAddClick(schema),
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      ) })
    ] })
  ] });
}

// src/SubmitButton/SubmitButton.tsx
var import_react15 = require("@chakra-ui/react");
var import_utils11 = require("@rjsf/utils");
var import_jsx_runtime13 = require("react/jsx-runtime");
function SubmitButton({ uiSchema }) {
  const { submitText, norender, props: submitButtonProps } = (0, import_utils11.getSubmitButtonOptions)(uiSchema);
  if (norender) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime13.jsx)(import_react15.Box, { marginTop: 3, children: /* @__PURE__ */ (0, import_jsx_runtime13.jsx)(import_react15.Button, { type: "submit", variant: "solid", ...submitButtonProps, children: submitText }) });
}

// src/TitleField/TitleField.tsx
var import_react16 = require("@chakra-ui/react");
var import_jsx_runtime14 = require("react/jsx-runtime");
function TitleField({
  id,
  title
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsxs)(import_react16.Box, { id, mt: 1, mb: 4, children: [
    /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_react16.Heading, { as: "h5", children: title }),
    /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_react16.Divider, {})
  ] });
}

// src/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx
var import_utils12 = require("@rjsf/utils");
var import_react17 = require("@chakra-ui/react");
var import_jsx_runtime15 = require("react/jsx-runtime");
function WrapIfAdditionalTemplate(props) {
  const {
    children,
    classNames,
    style,
    disabled,
    id,
    label,
    onDropPropertyClick,
    onKeyChange,
    readonly,
    registry,
    required,
    schema,
    uiSchema
  } = props;
  const { templates, translateString } = registry;
  const { RemoveButton: RemoveButton2 } = templates.ButtonTemplates;
  const keyLabel = translateString(import_utils12.TranslatableString.KeyLabel, [label]);
  const additional = import_utils12.ADDITIONAL_PROPERTY_FLAG in schema;
  if (!additional) {
    return /* @__PURE__ */ (0, import_jsx_runtime15.jsx)("div", { className: classNames, style, children });
  }
  const handleBlur = ({ target }) => onKeyChange(target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime15.jsxs)(import_react17.Grid, { className: classNames, style, alignItems: "center", gap: 2, children: [
    /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(import_react17.GridItem, { children: /* @__PURE__ */ (0, import_jsx_runtime15.jsxs)(import_react17.FormControl, { isRequired: required, children: [
      /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(import_react17.FormLabel, { htmlFor: `${id}-key`, id: `${id}-key-label`, children: keyLabel }),
      /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(
        import_react17.Input,
        {
          defaultValue: label,
          disabled: disabled || readonly,
          id: `${id}-key`,
          name: `${id}-key`,
          onBlur: !readonly ? handleBlur : void 0,
          type: "text",
          mb: 1
        }
      )
    ] }) }),
    /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(import_react17.GridItem, { children }),
    /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(import_react17.GridItem, { children: /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(
      RemoveButton2,
      {
        disabled: disabled || readonly,
        onClick: onDropPropertyClick(label),
        uiSchema,
        registry
      }
    ) })
  ] }, `${id}-key`);
}

// src/Templates/Templates.ts
function generateTemplates() {
  return {
    ArrayFieldItemTemplate,
    ArrayFieldTemplate,
    BaseInputTemplate,
    ButtonTemplates: {
      CopyButton,
      AddButton,
      MoveDownButton,
      MoveUpButton,
      RemoveButton,
      SubmitButton
    },
    DescriptionFieldTemplate: DescriptionField,
    ErrorListTemplate: ErrorList,
    FieldErrorTemplate,
    FieldHelpTemplate,
    FieldTemplate,
    ObjectFieldTemplate,
    TitleFieldTemplate: TitleField,
    WrapIfAdditionalTemplate
  };
}
var Templates_default = generateTemplates();

// src/AltDateWidget/AltDateWidget.tsx
var import_react18 = require("react");
var import_utils13 = require("@rjsf/utils");
var import_react19 = require("@chakra-ui/react");
var import_jsx_runtime16 = require("react/jsx-runtime");
function DateElement(props) {
  const { SelectWidget: SelectWidget2 } = props.registry.widgets;
  const value = props.value ? props.value : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
    SelectWidget2,
    {
      ...props,
      label: "",
      className: "form-control",
      onChange: (elemValue) => props.select(props.type, elemValue),
      options: {
        enumOptions: (0, import_utils13.dateRangeOptions)(props.range[0], props.range[1])
      },
      placeholder: props.type,
      schema: { type: "integer" },
      value,
      "aria-describedby": (0, import_utils13.ariaDescribedByIds)(props.name)
    }
  );
}
var readyForChange = (state) => {
  return Object.keys(state).every((key) => typeof state[key] !== "undefined" && state[key] !== -1);
};
function AltDateWidget(props) {
  const { autofocus, disabled, id, onBlur, onChange, onFocus, options, readonly, registry, showTime, value } = props;
  const { translateString } = registry;
  const [state, setState] = (0, import_react18.useState)((0, import_utils13.parseDateString)(value, showTime));
  (0, import_react18.useEffect)(() => {
    setState((0, import_utils13.parseDateString)(value, showTime));
  }, [showTime, value]);
  const handleChange = (property, nextValue) => {
    const nextState = {
      ...state,
      [property]: typeof nextValue === "undefined" ? -1 : nextValue
    };
    if (readyForChange(nextState)) {
      onChange((0, import_utils13.toDateString)(nextState, showTime));
    } else {
      setState(nextState);
    }
  };
  const handleNow = (event) => {
    event.preventDefault();
    if (disabled || readonly) {
      return;
    }
    const nextState = (0, import_utils13.parseDateString)((/* @__PURE__ */ new Date()).toJSON(), showTime);
    onChange((0, import_utils13.toDateString)(nextState, showTime));
  };
  const handleClear = (event) => {
    event.preventDefault();
    if (disabled || readonly) {
      return;
    }
    onChange(void 0);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime16.jsxs)(import_react19.Box, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(import_react19.Box, { display: "flex", flexWrap: "wrap", alignItems: "center", children: (0, import_utils13.getDateElementProps)(
      state,
      showTime,
      options.yearsRange,
      options.format
    ).map((elemProps, i) => {
      const elemId = id + "_" + elemProps.type;
      return /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(import_react19.Box, { mr: "2", mb: "2", children: /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
        DateElement,
        {
          ...props,
          ...elemProps,
          autofocus: autofocus && i === 0,
          disabled,
          id: elemId,
          name: id,
          onBlur,
          onFocus,
          readonly,
          registry,
          select: handleChange,
          value: elemProps.value < 0 ? "" : elemProps.value
        }
      ) }, elemId);
    }) }),
    /* @__PURE__ */ (0, import_jsx_runtime16.jsxs)(import_react19.Box, { display: "flex", children: [
      !options.hideNowButton && /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(import_react19.Button, { onClick: (e) => handleNow(e), mr: "2", children: translateString(import_utils13.TranslatableString.NowLabel) }),
      !options.hideClearButton && /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(import_react19.Button, { onClick: (e) => handleClear(e), children: translateString(import_utils13.TranslatableString.ClearLabel) })
    ] })
  ] });
}
AltDateWidget.defaultProps = {
  autofocus: false,
  disabled: false,
  readonly: false,
  showTime: false,
  options: {
    yearsRange: [1900, (/* @__PURE__ */ new Date()).getFullYear() + 2]
  }
};
var AltDateWidget_default = AltDateWidget;

// src/AltDateTimeWidget/AltDateTimeWidget.tsx
var import_jsx_runtime17 = require("react/jsx-runtime");
function AltDateTimeWidget(props) {
  const { AltDateWidget: AltDateWidget2 } = props.registry.widgets;
  return /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(AltDateWidget2, { ...props, showTime: true });
}
AltDateTimeWidget.defaultProps = {
  ...AltDateWidget_default.defaultProps,
  showTime: true
};
var AltDateTimeWidget_default = AltDateTimeWidget;

// src/CheckboxWidget/CheckboxWidget.tsx
var import_react20 = require("@chakra-ui/react");
var import_utils14 = require("@rjsf/utils");
var import_jsx_runtime18 = require("react/jsx-runtime");
function CheckboxWidget(props) {
  const {
    id,
    value,
    disabled,
    readonly,
    onChange,
    onBlur,
    onFocus,
    label,
    hideLabel,
    registry,
    options,
    uiSchema,
    schema
  } = props;
  const chakraProps = getChakra({ uiSchema });
  const required = (0, import_utils14.schemaRequiresTrueValue)(schema);
  const DescriptionFieldTemplate = (0, import_utils14.getTemplate)(
    "DescriptionFieldTemplate",
    registry,
    options
  );
  const description = options.description || schema.description;
  const _onChange = ({ target: { checked } }) => onChange(checked);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime18.jsxs)(import_react20.FormControl, { mb: 1, ...chakraProps, isRequired: required, children: [
    !hideLabel && !!description && /* @__PURE__ */ (0, import_jsx_runtime18.jsx)(
      DescriptionFieldTemplate,
      {
        id: (0, import_utils14.descriptionId)(id),
        description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime18.jsx)(
      import_react20.Checkbox,
      {
        id,
        name: id,
        isChecked: typeof value === "undefined" ? false : value,
        isDisabled: disabled || readonly,
        onChange: _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": (0, import_utils14.ariaDescribedByIds)(id),
        children: (0, import_utils14.labelValue)(/* @__PURE__ */ (0, import_jsx_runtime18.jsx)(import_react20.Text, { children: label }), hideLabel || !label)
      }
    )
  ] });
}

// src/CheckboxesWidget/CheckboxesWidget.tsx
var import_react21 = require("@chakra-ui/react");
var import_utils16 = require("@rjsf/utils");
var import_jsx_runtime19 = require("react/jsx-runtime");
function CheckboxesWidget(props) {
  const {
    id,
    disabled,
    options,
    value,
    readonly,
    onChange,
    onBlur,
    onFocus,
    required,
    label,
    hideLabel,
    uiSchema,
    rawErrors = []
  } = props;
  const { enumOptions, enumDisabled, emptyValue } = options;
  const chakraProps = getChakra({ uiSchema });
  const checkboxesValues = Array.isArray(value) ? value : [value];
  const _onBlur = ({ target }) => onBlur(id, (0, import_utils16.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, (0, import_utils16.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const row = options ? options.inline : false;
  const selectedIndexes = (0, import_utils16.enumOptionsIndexForValue)(value, enumOptions, true);
  return /* @__PURE__ */ (0, import_jsx_runtime19.jsxs)(
    import_react21.FormControl,
    {
      mb: 1,
      ...chakraProps,
      isDisabled: disabled || readonly,
      isRequired: required,
      isReadOnly: readonly,
      isInvalid: rawErrors && rawErrors.length > 0,
      children: [
        (0, import_utils16.labelValue)(
          /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(import_react21.FormLabel, { htmlFor: id, id: `${id}-label`, children: label }),
          hideLabel || !label
        ),
        /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(
          import_react21.CheckboxGroup,
          {
            onChange: (option) => onChange((0, import_utils16.enumOptionsValueForIndex)(option, enumOptions, emptyValue)),
            defaultValue: selectedIndexes,
            "aria-describedby": (0, import_utils16.ariaDescribedByIds)(id),
            children: /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(import_react21.Stack, { direction: row ? "row" : "column", children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
              const checked = (0, import_utils16.enumOptionsIsSelected)(option.value, checkboxesValues);
              const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
              return /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(
                import_react21.Checkbox,
                {
                  id: (0, import_utils16.optionId)(id, index),
                  name: id,
                  value: String(index),
                  isChecked: checked,
                  isDisabled: disabled || itemDisabled || readonly,
                  onBlur: _onBlur,
                  onFocus: _onFocus,
                  children: option.label && /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(import_react21.Text, { children: option.label })
                },
                index
              );
            }) })
          }
        )
      ]
    }
  );
}

// src/RadioWidget/RadioWidget.tsx
var import_react22 = require("@chakra-ui/react");
var import_utils18 = require("@rjsf/utils");
var import_jsx_runtime20 = require("react/jsx-runtime");
function RadioWidget({
  id,
  options,
  value,
  required,
  disabled,
  readonly,
  label,
  hideLabel,
  onChange,
  onBlur,
  onFocus,
  uiSchema
}) {
  const { enumOptions, enumDisabled, emptyValue } = options;
  const chakraProps = getChakra({ uiSchema });
  const _onChange = (nextValue) => onChange((0, import_utils18.enumOptionsValueForIndex)(nextValue, enumOptions, emptyValue));
  const _onBlur = ({ target: { value: value2 } }) => onBlur(id, (0, import_utils18.enumOptionsValueForIndex)(value2, enumOptions, emptyValue));
  const _onFocus = ({ target: { value: value2 } }) => onFocus(id, (0, import_utils18.enumOptionsValueForIndex)(value2, enumOptions, emptyValue));
  const row = options ? options.inline : false;
  const selectedIndex = (0, import_utils18.enumOptionsIndexForValue)(value, enumOptions) ?? null;
  return /* @__PURE__ */ (0, import_jsx_runtime20.jsxs)(import_react22.FormControl, { mb: 1, ...chakraProps, isDisabled: disabled || readonly, isRequired: required, isReadOnly: readonly, children: [
    (0, import_utils18.labelValue)(
      /* @__PURE__ */ (0, import_jsx_runtime20.jsx)(import_react22.FormLabel, { htmlFor: id, id: `${id}-label`, children: label }),
      hideLabel || !label
    ),
    /* @__PURE__ */ (0, import_jsx_runtime20.jsx)(
      import_react22.RadioGroup,
      {
        onChange: _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        value: selectedIndex,
        name: id,
        "aria-describedby": (0, import_utils18.ariaDescribedByIds)(id),
        children: /* @__PURE__ */ (0, import_jsx_runtime20.jsx)(import_react22.Stack, { direction: row ? "row" : "column", children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
          const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
          return /* @__PURE__ */ (0, import_jsx_runtime20.jsx)(
            import_react22.Radio,
            {
              value: String(index),
              id: (0, import_utils18.optionId)(id, index),
              disabled: disabled || itemDisabled || readonly,
              children: option.label
            },
            index
          );
        }) })
      }
    )
  ] });
}

// src/RangeWidget/RangeWidget.tsx
var import_react23 = require("@chakra-ui/react");
var import_utils20 = require("@rjsf/utils");
var import_jsx_runtime21 = require("react/jsx-runtime");
function RangeWidget({
  value,
  readonly,
  disabled,
  onBlur,
  onFocus,
  options,
  schema,
  uiSchema,
  onChange,
  label,
  hideLabel,
  id
}) {
  const chakraProps = getChakra({ uiSchema });
  const sliderWidgetProps = { value, label, id, ...(0, import_utils20.rangeSpec)(schema) };
  const _onChange = (value2) => onChange(value2 === void 0 ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_react23.FormControl, { mb: 1, ...chakraProps, children: [
    (0, import_utils20.labelValue)(/* @__PURE__ */ (0, import_jsx_runtime21.jsx)(import_react23.FormLabel, { htmlFor: id, children: label }), hideLabel || !label),
    /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(
      import_react23.Slider,
      {
        ...sliderWidgetProps,
        id,
        name: id,
        isDisabled: disabled || readonly,
        onChange: _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": (0, import_utils20.ariaDescribedByIds)(id),
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(import_react23.SliderTrack, { children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(import_react23.SliderFilledTrack, {}) }),
          /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(import_react23.SliderThumb, {})
        ]
      }
    )
  ] });
}

// src/SelectWidget/SelectWidget.tsx
var import_react24 = require("react");
var import_react25 = require("@chakra-ui/react");
var import_utils22 = require("@rjsf/utils");
var import_chakra_react_select = require("chakra-react-select");
var import_jsx_runtime22 = require("react/jsx-runtime");
function SelectWidget(props) {
  const {
    id,
    options,
    label,
    hideLabel,
    placeholder,
    multiple,
    required,
    disabled,
    readonly,
    value,
    autofocus,
    onChange,
    onBlur,
    onFocus,
    rawErrors = [],
    uiSchema,
    schema
  } = props;
  const { enumOptions, enumDisabled, emptyValue } = options;
  const chakraProps = getChakra({ uiSchema });
  const _onMultiChange = (e) => {
    return onChange(
      (0, import_utils22.enumOptionsValueForIndex)(
        e.map((v) => {
          return v.value;
        }),
        enumOptions,
        emptyValue
      )
    );
  };
  const _onChange = (e) => {
    return onChange((0, import_utils22.enumOptionsValueForIndex)(e.value, enumOptions, emptyValue));
  };
  const _onBlur = ({ target }) => onBlur(id, (0, import_utils22.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, (0, import_utils22.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const showPlaceholderOption = !multiple && schema.default === void 0;
  const { valueLabelMap, displayEnumOptions } = (0, import_react24.useMemo)(() => {
    const valueLabelMap2 = {};
    let displayEnumOptions2 = [];
    if (Array.isArray(enumOptions)) {
      displayEnumOptions2 = enumOptions.map((option, index) => {
        const { value: value2, label: label2 } = option;
        valueLabelMap2[index] = label2 || String(value2);
        return {
          label: label2,
          value: String(index),
          isDisabled: Array.isArray(enumDisabled) && enumDisabled.indexOf(value2) !== -1
        };
      });
      if (showPlaceholderOption) {
        displayEnumOptions2.unshift({ value: "", label: placeholder || "" });
      }
    }
    return { valueLabelMap: valueLabelMap2, displayEnumOptions: displayEnumOptions2 };
  }, [enumDisabled, enumOptions, placeholder, showPlaceholderOption]);
  const isMultiple = typeof multiple !== "undefined" && multiple !== false && Boolean(enumOptions);
  const selectedIndex = (0, import_utils22.enumOptionsIndexForValue)(value, enumOptions, isMultiple);
  const formValue = isMultiple ? (selectedIndex || []).map((i) => {
    return {
      label: valueLabelMap[i],
      value: i
    };
  }) : {
    label: valueLabelMap[selectedIndex] || "",
    selectedIndex
  };
  return /* @__PURE__ */ (0, import_jsx_runtime22.jsxs)(
    import_react25.FormControl,
    {
      mb: 1,
      ...chakraProps,
      isDisabled: disabled || readonly,
      isRequired: required,
      isReadOnly: readonly,
      isInvalid: rawErrors && rawErrors.length > 0,
      children: [
        (0, import_utils22.labelValue)(
          /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(import_react25.FormLabel, { htmlFor: id, id: `${id}-label`, children: label }),
          hideLabel || !label
        ),
        /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(
          import_chakra_react_select.Select,
          {
            inputId: id,
            name: id,
            isMulti: isMultiple,
            options: displayEnumOptions,
            placeholder,
            closeMenuOnSelect: !isMultiple,
            onBlur: _onBlur,
            onChange: isMultiple ? _onMultiChange : _onChange,
            onFocus: _onFocus,
            autoFocus: autofocus,
            value: formValue,
            "aria-describedby": (0, import_utils22.ariaDescribedByIds)(id)
          }
        )
      ]
    }
  );
}

// src/TextareaWidget/TextareaWidget.tsx
var import_react26 = require("@chakra-ui/react");
var import_utils24 = require("@rjsf/utils");
var import_jsx_runtime23 = require("react/jsx-runtime");
function TextareaWidget({
  id,
  placeholder,
  value,
  label,
  hideLabel,
  disabled,
  autofocus,
  readonly,
  onBlur,
  onFocus,
  onChange,
  options,
  uiSchema,
  required,
  rawErrors
}) {
  const chakraProps = getChakra({ uiSchema });
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime23.jsxs)(
    import_react26.FormControl,
    {
      mb: 1,
      ...chakraProps,
      isDisabled: disabled || readonly,
      isRequired: required,
      isReadOnly: readonly,
      isInvalid: rawErrors && rawErrors.length > 0,
      children: [
        (0, import_utils24.labelValue)(/* @__PURE__ */ (0, import_jsx_runtime23.jsx)(import_react26.FormLabel, { htmlFor: id, children: label }), hideLabel || !label),
        /* @__PURE__ */ (0, import_jsx_runtime23.jsx)(
          import_react26.Textarea,
          {
            id,
            name: id,
            value: value ?? "",
            placeholder,
            autoFocus: autofocus,
            onChange: _onChange,
            onBlur: _onBlur,
            onFocus: _onFocus,
            rows: options.rows,
            "aria-describedby": (0, import_utils24.ariaDescribedByIds)(id)
          }
        )
      ]
    }
  );
}

// src/UpDownWidget/UpDownWidget.tsx
var import_react27 = require("@chakra-ui/react");
var import_utils26 = require("@rjsf/utils");
var import_jsx_runtime24 = require("react/jsx-runtime");
function UpDownWidget(props) {
  const { id, uiSchema, readonly, disabled, label, hideLabel, value, onChange, onBlur, onFocus, rawErrors, required } = props;
  const chakraProps = getChakra({ uiSchema });
  const _onChange = (value2) => onChange(value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)(
    import_react27.FormControl,
    {
      mb: 1,
      ...chakraProps,
      isDisabled: disabled || readonly,
      isRequired: required,
      isReadOnly: readonly,
      isInvalid: rawErrors && rawErrors.length > 0,
      children: [
        (0, import_utils26.labelValue)(/* @__PURE__ */ (0, import_jsx_runtime24.jsx)(import_react27.FormLabel, { htmlFor: id, children: label }), hideLabel || !label),
        /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)(
          import_react27.NumberInput,
          {
            value: value ?? "",
            onChange: _onChange,
            onBlur: _onBlur,
            onFocus: _onFocus,
            "aria-describedby": (0, import_utils26.ariaDescribedByIds)(id),
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(import_react27.NumberInputField, { id, name: id }),
              /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)(import_react27.NumberInputStepper, { children: [
                /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(import_react27.NumberIncrementStepper, {}),
                /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(import_react27.NumberDecrementStepper, {})
              ] })
            ]
          }
        )
      ]
    }
  );
}

// src/Widgets/Widgets.ts
function generateWidgets() {
  return {
    AltDateTimeWidget: AltDateTimeWidget_default,
    AltDateWidget: AltDateWidget_default,
    CheckboxWidget,
    CheckboxesWidget,
    RadioWidget,
    RangeWidget,
    SelectWidget,
    TextareaWidget,
    UpDownWidget
  };
}
var Widgets_default = generateWidgets();

// src/Theme/Theme.tsx
function generateTheme() {
  return {
    templates: generateTemplates(),
    widgets: generateWidgets()
  };
}
var Theme_default = generateTheme();

// src/Form/Form.tsx
function generateForm() {
  return (0, import_core.withTheme)(generateTheme());
}
var Form_default = generateForm();

// src/ChakraFrameProvider.tsx
var import_react29 = require("@emotion/react");
var import_cache = __toESM(require("@emotion/cache"));
var import_weak_memoize = __toESM(require("@emotion/weak-memoize"));
var import_react30 = require("@chakra-ui/react");

// src/CssReset.tsx
var import_react28 = require("@emotion/react");
var import_jsx_runtime25 = require("react/jsx-runtime");
var CSSReset = () => /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(
  import_react28.Global,
  {
    styles: `
      html {
        line-height: 1.5;
        -webkit-text-size-adjust: 100%;
        font-family: system-ui, sans-serif;
        -webkit-font-smoothing: antialiased;
        text-rendering: optimizeLegibility;
        -moz-osx-font-smoothing: grayscale;
        touch-action: manipulation;
      }
      body {
        position: relative;
        min-height: 100%;
        font-feature-settings: 'kern';
      }
      *,
      *::before,
      *::after {
        border-width: 0;
        border-style: solid;
        box-sizing: border-box;
      }
      main {
        display: block;
      }
      hr {
        border-top-width: 1px;
        box-sizing: content-box;
        height: 0;
        overflow: visible;
      }
      pre,
      code,
      kbd,
      samp {
        font-family: SFMono-Regular,  Menlo, Monaco, Consolas, monospace;
        font-size: 1em;
      }
      a {
        background-color: transparent;
        color: inherit;
        text-decoration: inherit;
      }
      abbr[title] {
        border-bottom: none;
        text-decoration: underline;
        -webkit-text-decoration: underline dotted;
        text-decoration: underline dotted;
      }
      b,
      strong {
        font-weight: bold;
      }
      small {
        font-size: 80%;
      }
      sub,
      sup {
        font-size: 75%;
        line-height: 0;
        position: relative;
        vertical-align: baseline;
      }
      sub {
        bottom: -0.25em;
      }
      sup {
        top: -0.5em;
      }
      img {
        border-style: none;
      }
      button,
      input,
      optgroup,
      select,
      textarea {
        font-family: inherit;
        font-size: 100%;
        line-height: 1.15;
        margin: 0;
      }
      button,
      input {
        overflow: visible;
      }
      button,
      select {
        text-transform: none;
      }
      button::-moz-focus-inner,
      [type="button"]::-moz-focus-inner,
      [type="reset"]::-moz-focus-inner,
      [type="submit"]::-moz-focus-inner {
        border-style: none;
        padding: 0;
      }
      fieldset {
        padding: 0.35em 0.75em 0.625em;
      }
      legend {
        box-sizing: border-box;
        color: inherit;
        display: table;
        max-width: 100%;
        padding: 0;
        white-space: normal;
      }
      progress {
        vertical-align: baseline;
      }
      textarea {
        overflow: auto;
      }
      [type="checkbox"],
      [type="radio"] {
        box-sizing: border-box;
        padding: 0;
      }
      [type="number"]::-webkit-inner-spin-button,
      [type="number"]::-webkit-outer-spin-button {
        -webkit-appearance: none !important;
      }
      input[type="number"] {
        -moz-appearance: textfield;
      }
      [type="search"] {
        -webkit-appearance: textfield;
        outline-offset: -2px;
      }
      [type="search"]::-webkit-search-decoration {
        -webkit-appearance: none !important;
      }
      ::-webkit-file-upload-button {
        -webkit-appearance: button;
        font: inherit;
      }
      details {
        display: block;
      }
      summary {
        display: list-item;
      }
      template {
        display: none;
      }
      [hidden] {
        display: none !important;
      }
      body,
      blockquote,
      dl,
      dd,
      h1,
      h2,
      h3,
      h4,
      h5,
      h6,
      hr,
      figure,
      p,
      pre {
        margin: 0;
      }
      button {
        background: transparent;
        padding: 0;
      }
      fieldset {
        margin: 0;
        padding: 0;
      }
      ol,
      ul {
        margin: 0;
        padding: 0;
      }
      textarea {
        resize: vertical;
      }
      button,
      [role="button"] {
        cursor: pointer;
      }
      button::-moz-focus-inner {
        border: 0 !important;
      }
      table {
        border-collapse: collapse;
      }
      h1,
      h2,
      h3,
      h4,
      h5,
      h6 {
        font-size: inherit;
        font-weight: inherit;
      }
      button,
      input,
      optgroup,
      select,
      textarea {
        padding: 0;
        line-height: inherit;
        color: inherit;
      }
      img,
      svg,
      video,
      canvas,
      audio,
      iframe,
      embed,
      object {
        display: block;
      }
      img,
      video {
        max-width: 100%;
        height: auto;
      }
      [data-js-focus-visible] :focus:not([data-focus-visible-added]) {
        outline: none;
        box-shadow: none;
      }
      select::-ms-expand {
        display: none;
      }
      input {
        border-width: revert;
        border-color: revert;
        border-style: revert;
      }
      .array-item > hr {
        margin-top: 16px;
        margin-bottom: 16px;
      }
    `
  }
);
var CssReset_default = CSSReset;

// src/ChakraFrameProvider.tsx
var import_jsx_runtime26 = require("react/jsx-runtime");
var memoizedCreateCacheWithContainer = (0, import_weak_memoize.default)((container) => {
  const newCache = (0, import_cache.default)({ container, key: "rjsf" });
  return newCache;
});
var __createChakraFrameProvider = (props) => ({ document }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime26.jsx)("div", { style: { margin: 2 }, children: /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(import_react29.CacheProvider, { value: memoizedCreateCacheWithContainer(document.head), children: /* @__PURE__ */ (0, import_jsx_runtime26.jsxs)(import_react30.ChakraProvider, { resetCSS: false, children: [
    /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(CssReset_default, {}),
    props.children
  ] }) }) });
};

// src/index.ts
var src_default = Form_default;
